DataOnion/Auth/TwoFactorRedisContext.cs

```diff
@@ -0,0 +1,176 @@
+            string userId,
+            VerificationMethod method
+        );
+        Task DeleteTwoFactorRequestsAsync(
+            string userId
+        );
+    }
+
+    public class TwoFactorRequest
+    {
+        public string Did { get; set; }
+        public int VerificationCode { get; set; }
+        public DateTime CreatedAt { get; set; }
+        public Guid Token { get; set; }
+        public VerificationMethod Method { get; set; }
+    }
+
+    public class TwoFactorRedisContext : ITwoFactorRedisContext
+    {
+        private readonly string _environmentPrefix;
+        private readonly string _twoFactorRequestPrefix;
+        private readonly ILogger? _logger;
+        private readonly IRedisManager _redisManager;
+        private IDatabase _database => _redisManager.GetDatabase();
+
+        public TwoFactorRedisContext(
+            string environmentPrefix,
+            string twoFactorRequestPrefix,
+            IRedisManager redisManager,
+            ILogger<TwoFactorRedisContext>? logger
+        )
+        {
+            _environmentPrefix = environmentPrefix;
+            _twoFactorRequestPrefix = twoFactorRequestPrefix;
+            _redisManager = redisManager;
+            _logger = logger;
+        }
+
+        public async Task CreateTwoFactorRequestAsync(string userId, TwoFactorRequest request)
+        {
+            var keyPrefix = _twoFactorRequestPrefix;
+            var method = request.Method;
+            var redisKey = BuildTwoFactorKey(
+                userId,
+                method
+            );
+            var stringEnumConverter = new JsonStringEnumConverter();
+            var serializationOptions = new JsonSerializerOptions();
+            serializationOptions.Converters.Add(stringEnumConverter);
+
+            await _database.ListLeftPushAsync(
+                redisKey, 
+                new RedisValue(
+                    JsonSerializer.Serialize(
+                        request, 
+                        serializationOptions
+                    )
+                )
+            );
+        }
+
+        public async Task<IEnumerable<TwoFactorRequest>> FetchTwoFactorRequestsAsync(
+            string userId,
+            VerificationMethod method
+        )
+        {
+            var key = BuildTwoFactorKey(userId, method);
+
+            try
+            {
+                var stringEnumConverter = new JsonStringEnumConverter();
+                var serializationOptions = new JsonSerializerOptions();
+                serializationOptions.Converters.Add(stringEnumConverter);
+
+                var values = await FetchValuesAsync(
+                    key
+                );
+
+                var requests = values.Select(val => 
+                    JsonSerializer.Deserialize<TwoFactorRequest>(
+                        val, 
+                        serializationOptions
+                    )
+                );
+
+                return requests;
```

Comment:
Not sure how much we care about this, but I'm getting a nullability warning here. `JsonSerializer.Deserialize` returns a nullable object, so technically this is an `IEnumerable<TwoFactorRequest?>` rather than an `IEnumerable<TwoFactorRequest>`.

According to one source I found, you can use `.OfType<T>()` to remove all the nulls from an `IEnumerable<T?>`, which would make the above statement `values.Select(...).OfType<TwoFactorRequest>()`. I'm not sure if that's the best way to solve this, though.

Reply:
I tried to use `.Where()`, and that didn't work since it filtered by predicate, not by type. I think `.OfType<T>()` is our best option here.